server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -345,9 +347,33 @@ private <C extends Collector, T> T search(Weight weight, CollectorManager<C, T>
 
     @Override
     public void search(List<LeafReaderContext> leaves, Weight weight, Collector collector) throws IOException {
-        collector.setWeight(weight);
-        for (LeafReaderContext ctx : leaves) { // search each subreader
-            searchLeaf(ctx, weight, collector);
+        try {
+            collector.setWeight(weight);
+            for (LeafReaderContext ctx : leaves) { // search each subreader
+                searchLeaf(ctx, weight, collector);
+            }
+        } catch (@SuppressWarnings("unused") TimeExceededException e) {
+            // ideally we should just throw an error here if we don't want partial results
+            timeExceeded = true;
+        }
+
+    }
+
+    /**  If the search has timed out following Elasticsearch custom implementation */
+    public boolean timeExceeded() {
+        return timeExceeded;
+    }
+
+    public RuntimeException getTimeExceededException() {
+        return new TimeExceededException();
```

Comment:
make it throw directly without returning it to then throw it?

Reply:
done